text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,397 @@
+OpenTelemetry should provide a solution to end users that allows them to provide
+a configuration for their redaction requirements. This configuration consists of
+two pieces:
+
+* Redaction requirement "configuration profiles", that end user use to express
+  their overall redaction requirements.
+* Fine grained configuration that can be used to express requirements that are
+  unique to their environment.
+
+**Example**: An end user may configure an application with "stricter" requirements
+ and some custom requirements like the following:
+
+* They provide a profile name `STRICTER` via an environment variable, e.g.
+  OTEL_SENSITIVE_DATA_PROFILE="STRICTER"
+* They add custom requirements in a
+  [configuration file](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md), e.g.
+  
+  ```yaml
+  sensitiveData:
+      - attribute: url.query
+        redaction: REDACT_ALL_URL_VALUES
+      - attribute: com.example.customer.name
+        redaction: DROP
+      - attribute: com.example.customer.email
+        redaction: 's/^[^@]*@/REDACTED@/'
+  ```
+
+The above is just an example for how such a configuration may look like, different
+formats are discussed below.
+
+#### Redaction requirement profiles
+
+An end user should be able to pick from a set of redaction requirement profiles
+that cover common configurations:
+
+- `NONE` or `NEVER`: No redaction is applied. This is useful if an end user
+  chooses to run a collector nearby the application that will take care of
+  redaction out of process. If this profile is selected the SDK must make sure
+  that (almost) no overhead remains from the redaction methods.
+- `DEFAULT`: A set of rules that especially make sure that no security-sensitive
+  data is exposed, only limited PII-related redactions are applied.
+- `STRICT`: A set that builds on top of `DEFAULT` and also applies commonly required
+  PII-related redactions (anonymize emails, names, IPs, etc)
+  `STRICTER`: A set that builds on top of `STRICT` and applies even more redactions,
+  e.g. drop all query attribute values, drop PII-related data, etc.
+- `DROP` or `ALWAYS`: All values with (potentially) sensitivity details will be dropped
+
+Those profiles will be defined through the semantic conventions (see below).
+
+#### Fine-grained configuration
+
+On top of the redaction requirements profiles an end user should have capabilities
+to add their own local requirements, either through configuration or through code.
+
+The configuration is a list of rules that the SDK can parse and apply. An item
+in this list will consist of:
+
+- Conditions when a redaction should be applied
+- Instructions for the redaction that will be applied when the conditions are met
+
+In an example configuration file a list could look like the following:
+
+```yaml
+  sensitiveData:
+      - attribute: url.query
+        redaction: REDACT_ALL_URL_VALUES
```

Comment:
Do we need to only specify "redaction" that can be implemented in all languages?    Will this be a new specification of allowable redactions?

Reply:
maybe? I think at some point we need to do some prototyping to figure out what redaction is allowed to do and what not. Performance is the main reason but "this needs to be implemented in "any" language" is another complex requirement.